### ISSUE TITLE
Configuration for rule "comma-dangle" is invalid

### ISSUE THREAD

**Issue by sethen** [eslint/eslint#2771]
Having an issue with my `.eslintrc` in the root of my project.  Using `eslint 0.23.0` and `node 0.12.2` -- here are my rules:

```
{
    "rules": {
        "comma-dangle": "never",
        "no-cond-assign": "always",
        "no-console": 1,
        "no-constant-condition": 1,
        "no-control-regex": 0,
        "no-debugger": 1,
        "no-dupe-args": 1,
        "no-dupe-keys": 1,
        "no-duplicate-case": 1,
        "no-empty": 1,
        "no-ex-assign": 1,
        "no-extra-boolean-cast": 0,
        "no-extra-parens": 0,
        "no-extra-semi": 1,
        "no-func-assign": 1,
        "no-inner-declarations": 1,
        "no-invalid-regexp": 1,
        "no-irregular-whitespace": 1,
        "no-negated-in-lhs": 1,
        "no-obj-calls": 1,
        "no-regex-spaces": 1,
        "no-reserved-keys": 1,
        "no-sparse-arrays": 1,
        "no-unreachable": 1,
        "use-isnan": 1,
        "valid-jsdoc": 0,
        "valid-typeof": 1,

        "block-scoped-var": 0,
        "complexity": [1, 2],
        "consistent-return": 1,
        "curly": [2, "all"],
        "default-case": 1,
        "dot-notation": 1,
        "dot-location": "object",
        "eqeqeq": [2, "smart"],
        "guard-for-in": 1,
        "no-alert": 1,
        "no-caller": 1,
        "no-div-regex": 1,
        "no-else-return": 1,
        "no-empty-label": 0,
        "no-eq-null": 1,
        "no-eval": 1,
        "no-extend-native": 1,
        "no-extra-bind": 1,
        "no-fallthrough": 1,
        "no-floating-decimal": 1,
        "no-implied-eval": 1,
        "no-iterator": 1,
        "no-labels": 1,
        "no-lone-blocks": 0,
        "no-loop-func": 1,
        "no-multi-spaces": 1,
        "no-multi-str": 0,
        "no-native-reassign": 1,
        "no-new-func": 1,
        "no-new-wrappers": 1,
        "no-new": 1,
        "no-octal-escape": 1,
        "no-octal": 1,
        "no-param-reassign": 1,
        "no-process-env": 0,
        "no-proto": 1,
        "no-redeclare": 1,
        "no-return-assign": 1,
        "no-script-url": 1,
        "no-self-compare": 1,
        "no-sequences": 1,
        "no-throw-literal": 1,
        "no-unused-expressions": 1,
        "no-void": 1,
        "no-with": 1,
        "radix": 1,
        "vars-on-top": 0,
        "wrap-iife": 1,
        "yoda": 0,

        "no-catch-shadow": 1,
        "no-delete-var": 0,
        "no-label-var": 1,
        "no-shadow-restricted-names": 1,
        "no-shadow": 1,
        "no-undef-init": 1,
        "no-undef": 1,
        "no-undefined": 1,
        "no-unused-vars": 1,
        "no-use-before-define": 1,

        "brace-style": [2, "stroustrup"],
        "camelcase": [2, {"properties": "always"}],
        "comma-spacing": [2, {"before": false, "after": true}],
        "comma-style": [2, "last"],
        "consistent-this": [0, "self"],
        "eol-last": 0,
        "func-names": 0,
        "func-style": [2, "declaration"],
        "indent": 2,
        "key-spacing": [2, {
            "beforeColon": false,
            "afterColon": true
        }],
        "linebreak-style": "unix",
        "max-nested-callbacks": [2, 3],
        "new-cap": 1,
        "new-parens": 1,
        "newline-after-var": 1,
        "no-array-constructor": 0,
        "no-continue": 0,
        "no-inline-comments": 0,
        "no-lonely-if": 1,
        "no-mixed-spaces-and-tabs": 1,
        "no-multiple-empty-lines": [1, {max: 2}],
        "no-nested-ternary": 1,
        "no-new-object": 1,
        "no-spaced-func": 1,
        "no-ternary": 0,
        "no-trailing-spaces": 1,
        "no-underscore-dangle": 0,
        "no-unneeded-ternary": 1,
        "no-wrap-func": 1,
            "one-var": 0,
        "operator-assignment": [2, "always"],
        "operator-linebreak": [2, "before"],
        "quote-props": 0,
        "quotes": [2, "single", "avoid-escape"],
        "semi-spacing": [2, {"before": false, "after": true}],
        "semi": [2, "always"],
        "space-after-keywords": "always",
        "space-before-blocks": "always",
        "space-before-function-paren": {"anonymous": "always", "named": "never"},
        "space-in-brackets": "always",
        "space-infix-ops": true,
        "space-return-throw-case": 1,
        "space-unary-ops": [1, { "words": true, "nonwords": false }],
        "spaced-line-comment": "always",
        "wrap-regex": 0,

        // ES6 RULES

        "generator-star-spacing": [2, "before"],
        "no-var": 0,
        "object-shorthand": 1,

        // REACT

        "react/display-name": 0,
        "react/jsx-boolean-value": "never",
        "react/jsx-no-undef": 1,
        "react/jsx-quotes": [2, "single"],
        "react/jsx-sort-prop-types": 0,
        "react/jsx-sort-props": 0,
        "react/jsx-uses-react": 1,
        "react/jsx-uses-vars": 1,
        "react/no-did-mount-set-state": 0,
        "react/no-did-update-set-state": 0,
        "react/no-multi-comp": 0,
        "react/no-unknown-property": 1,
        "react/prop-types": 1,
        "react/react-in-jsx-scope": 1,
        "react/self-closing-comp": 1,
        "react/sort-comp": [1, {
            "order": [
                "lifecycle",
                "render"
            ],
            "groups": {
                "lifecycle": [
                    "displayName",
                    "propTypes",
                    "contextTypes",
                    "childContextTypes",
                    "mixins",
                    "statics",
                    "getDefaultProps",
                    "getInitialState",
                    "getChildContext",
                    "componentWillMount",
                    "componentDidMount",
                    "componentWillReceiveProps",
                    "shouldComponentUpdate",
                    "componentWillUpdate",
                    "componentDidUpdate",
                    "componentWillUnmount"
                ]
            }
        }],
        "react/wrap-multilines": [1, {declaration: true, assignment: true, return: true}]
    },

    "ecmaFeatures": {
        "arrowFunctions": 1,
        "binaryLiterals": 1,
        "blockBindings": 1,
        "classes": 1,
        "defaultParams": 1,
        "destructuring": 1,
        "forOf": 1,
        "generators": 1,
        "modules": 1,
        "objectLiteralComputedProperties": 1,
        "objectLiteralDuplicateProperties": 1,
        "objectLiteralShorthandMethods": 1,
        "objectLiteralShorthandProperties": 1,
        "octalLiterals": 1,
        "regexUFlag": 1,
        "regexYFlag": 1,
        "restParams": 1,
        "spread": 1,
        "superInFunctions": 1,
        "templateStrings": 1,
        "unicodeCodePointEscapes": 1,
        "globalReturn": 1,
        "jsx": 1
    },

    "plugins": [
        "react"
    ]
}
```

Here's the error I am getting:

```
/usr/local/lib/node_modules/eslint/lib/config-validator.js:88
        throw new Error(message.join(""));
              ^
Error: /Users/sethen/Developer/insightpool-client/.eslintrc:
	Configuration for rule "comma-dangle" is invalid:
	Value "never" is the wrong type.

    at validateRuleOptions (/usr/local/lib/node_modules/eslint/lib/config-validator.js:88:15)
    at /usr/local/lib/node_modules/eslint/lib/config-validator.js:101:13
    at Array.forEach (native)
    at Object.validate (/usr/local/lib/node_modules/eslint/lib/config-validator.js:100:35)
    at loadConfig (/usr/local/lib/node_modules/eslint/lib/config.js:99:19)
    at getLocalConfig (/usr/local/lib/node_modules/eslint/lib/config.js:225:23)
    at Config.getConfig (/usr/local/lib/node_modules/eslint/lib/config.js:363:22)
    at processText (/usr/local/lib/node_modules/eslint/lib/cli-engine.js:183:27)
    at processFile (/usr/local/lib/node_modules/eslint/lib/cli-engine.js:226:12)
    at /usr/local/lib/node_modules/eslint/lib/cli-engine.js:322:26
```

These rules were working fine in `0.21.0` so I am not sure what exactly changed to make these rules invalid.

**Comment by gyandeeps** [comment]
You configuration is incorrect for many rules.
here is how the rule are configured.
<rule-name>: [<0 or 1 or 2>, <options>]

example
comma-dangle: [2, "never"]

**Comment by sethen** [comment]
@gyandeeps Ah, thanks.  Let me try this out and see if it resolves the issues I am having.

**Comment by sethen** [comment]
@gyandeeps This did fix my issue... Has this been cited anywhere in the release notes??  As I said, my rules we're working fine up until this `0.23.0` release.

**Comment by gyandeeps** [comment]
why it was working fine because it was simply ignoring that rule.

**Comment by sethen** [comment]
@gyandeeps Thanks a lot!

**Comment by ilyavolodin** [comment]
@sethen it was working because misconfigured rules were ignored. In 0.23.0 we added config validation, so now you are seeing errors for misconfigured rules,

**Comment by hax** [comment]
This is a breaking changes! It enforce many people to rewrite their config, especially add many redundant for yaml format config.

For example, my config
```yaml
  rules:
    semi: never
    dot-notation: 1
    quotes: single avoid-escape
``` 
need to rewrite as
```yaml
  rules:
    semi: [2, "never"]
    dot-notation: [1]
    quotes: [2, "single", "avoid-escape"]
```

I lose most benefit of using yaml!!

**Comment by ilyavolodin** [comment]
@hax I think you are misunderstanding the issue. With the config you posted above, eslint never checked for quotes or semi. Those rules were just ignored, because eslint couldn't understand configuration values that you provided. Now instead of blindly ignoring rules that are not configured correctly, it will notify you about them. I'm also not sure how adding array values makes you lose benefit of using yaml. Yaml doesn't magically add missing values for you, if the rule requires an array, and you are passing it a string, yaml isn't going to fix that for you. You can rewrite your config to the following:

```yaml
rules:
    semi:
        - 2
        - "never"
    dot-notation: 1
    quotes:
        - 2
        - "single"
        - "avoid-escape"
```

**Comment by hax** [comment]
@ilyavolodin Thank you, my misunderstanding :(

**Comment by dfmartin** [comment]
@ilyavolodin how would I configure quotes to allow template literals?

``` yml
rules:        
      quotes: 
            - 1
            - "single"
            - "allowtemplateliterals"
```
I have tried "allow-template-literals" and various other permutations, but I keep receiving: Value "allowTemplateLiterals" no schema.  Much appreciated.

**Comment by ilyavolodin** [comment]
@dfmartin Please join gitter chat if you have question about configuring ESLint. Old closed issues are not correct place for this type of questions.
